Project 1 part 2 Samuel Andrews


Reading in packages

In [57]:
import re

Following the plan laid out in triage, the first step will be to read in the file completely and turn it into a list
seperated by lines

In [58]:
#read in the file

with open('Sell Short Trades.txt') as f:
    content = f.read()
content[:5000]

'                                                                                                                \n DOTC                                                                                 RUN DATE : 06/14/17 21:56 \n                                                                                             RPT DATE :06/14/17 \n                                    TTS0126:SELL SHORT TRADES & SHORT EXEMPT                                    \n Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr \n                                                                                                                \n TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr      \n CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS        \n                                                                    /17  /17  256945   

In [59]:
#make it a list with each item being one line

grouped_lines = content.split('\n')
grouped_lines[:20]

['                                                                                                                ',
 ' DOTC                                                                                 RUN DATE : 06/14/17 21:56 ',
 '                                                                                             RPT DATE :06/14/17 ',
 '                                    TTS0126:SELL SHORT TRADES & SHORT EXEMPT                                    ',
 ' Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr ',
 '                                                                                                                ',
 ' TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr      ',
 ' CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS        ',
 '                                                              

Next, we will remove the starting and ending lines that do not provide meaningful content, checking first to ensure we do not remove any actual data

In [60]:
#look at just the top

grouped_lines[0:5]

['                                                                                                                ',
 ' DOTC                                                                                 RUN DATE : 06/14/17 21:56 ',
 '                                                                                             RPT DATE :06/14/17 ',
 '                                    TTS0126:SELL SHORT TRADES & SHORT EXEMPT                                    ',
 ' Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr ']

In [61]:
grouped_lines[4:10]

[' Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr ',
 '                                                                                                                ',
 ' TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr      ',
 ' CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS        ',
 '                                                                    /17  /17  256945       00                   ',
 ' 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO                                             ']

In [62]:
#remove those lines

grouped_lines = grouped_lines[4:]

grouped_lines[0:20]

[' Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr ',
 '                                                                                                                ',
 ' TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr      ',
 ' CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS        ',
 '                                                                    /17  /17  256945       00                   ',
 ' 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO                                             ',
 ' CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX  ',
 '                                                                    /17  /17  866087       04                   ',
 ' 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA

In [63]:
#During the triage phase, we determined everything after line 4,115 can be removed outright
#obiviously we will need to adjust for indexing and the removed top


print(grouped_lines[4110]) #last line we care about
print(grouped_lines[4111]) #garbage starts


                                             854300                                                             
                                               TOTAL TRADES: 913                                                


In [64]:
#removing the ending lines

grouped_lines = grouped_lines[0:4111]

grouped_lines[-1] #check the last line


'                                             854300                                                             '

In [65]:
#now that step 1 is complete, save to a different name for step 2

grouped_lines2 = grouped_lines

As described, for our second step we will try to further prune useless lines from the list. Starting out, there are numerous blank lines thrown in at various points, and we want to deal with all of them with one step. By making use of the strip function across the entire list, we can quickly make lines devoid of content length zero, which will allow a second pass through of the list to quickly identify the offending items and purge them thusly. 

In [66]:
#make use of a list comprehension to quickly iterate strip over all of the lines


grouped_lines2 = [x.strip() for x in grouped_lines2]

grouped_lines2[0:10] #notice line 2 in this example

['Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 '',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr',
 'CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300']

In [67]:
#as shown, lines like line 2 (index 1) are now length 0

len(grouped_lines2[1])

0

In [68]:
#building a lambda function to test the implementation

blank_marker = lambda line: len(line) == 0

#seen working here
print(blank_marker(grouped_lines2[1]))

print(blank_marker(grouped_lines2[0]))


True
False


In [69]:
#using a list comprehension to remove based on that comprehension (actually making a new list to be techincal)


test = [line for line in grouped_lines2 if len(line) != 0] #using a new variable to make sure it works properly

test[0:10] #notice line two's removal


['Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr',
 'CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX']

In [70]:
#examine test to ensure it worked as intended, you can see below where we hit the page break that used to have empty lines, now all removed

test[0:50]



['Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr',
 'CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866096       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          100   2.350000    2.350000   2.36000006/1406/191706149900003 10:20:ContraSSFREX',


In [71]:
#now confident, we can overwrite grouplines2

grouped_lines2 = test

With the blank lines removed, we can now deal with the page break blocks. Again, as outlined in the Triage step, the only instances of the word "of" within this entire document is on page break lines, which follow the form "x of 77" where x is any number. This is easily confirmed with a simple CTRL-F in notepad. therefore, we can use fundamentally the same solution as the previous step to remove these lines as well with us instead looking for lines containing "of".

In [72]:
#sample offending line

grouped_lines2[43]

'1 of 77'

In [73]:
#viewed in context
grouped_lines2[40:50]

['CERS     SS      NEW          100   2.360000    2.350000   2.36000006/1406/191706149900003 10:24:ContraSSFREX',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS /17  /17  909154       02',
 '854300',
 '1 of 77',
 'Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr',
 'CERS     SS      NEW          100   2.360000    2.350000   2.36000006/1406/191706149900003 10:24:ContraSSFREX',
 '/17  /17  909168       02',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300']

In [74]:

#building a lambda function to test the implementation

of_course = lambda line: "of" in line

#seen working here
print(of_course(grouped_lines2[43]))

print(of_course(grouped_lines2[44]))

True
False


In [75]:
test = [line for line in grouped_lines2 if "of" not in line] #using a new variable to make sure it works properly

test[40:50] #the page number's removal

['CERS     SS      NEW          100   2.360000    2.350000   2.36000006/1406/191706149900003 10:24:ContraSSFREX',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS /17  /17  909154       02',
 '854300',
 'Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr',
 'CERS     SS      NEW          100   2.360000    2.350000   2.36000006/1406/191706149900003 10:24:ContraSSFREX',
 '/17  /17  909168       02',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          100   2.360000    2.350000   2.36000006/1406/191706149900003 10:24:ContraSSFREX']

In [76]:
#setting up for the next part

grouped_lines2 = test

grouped_lines3 = grouped_lines2

At present, every row currently contained is either data or redudant columns contating column names or additional info. As part of step 3, we will be removing these problem rows. Foruntatley, as described in the triage step, due to our strippping, the computer-generated nature of the document, and the data lines themselves never sharing lines, we can use a similar solution to the previous step, intead searching for the two possible starters that begin these lines (Symbol and TradeCommType)

In [77]:
#example lines here (note the first two)

grouped_lines3[0:20]

['Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr',
 'CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866096       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          100   2.350000    2.350000   2.36000006/1406/191706149900003 10:20:ContraSSFREX',


In [78]:
#viewed on their own
grouped_lines3[0:2]

['Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'TradeCommType  SourceCommission   Account   OrderID      GTL               Trailer Info               Clr']

In [79]:
#building a lambda function to test the implementation

Symbol_start = lambda line: line.startswith("Symbol")

#seen working here
print(Symbol_start(grouped_lines3[0]))

print(Symbol_start(grouped_lines3[5]))

True
False


In [80]:
#looking at the same thing with TradeCommType

TradeCommType = lambda line: line.startswith("TradeCommType")

#seen working here
print(TradeCommType(grouped_lines3[1]))

print(TradeCommType(grouped_lines3[5]))

True
False


In [81]:
#using a new variable to make sure it works properly

test = grouped_lines3

test = [line for line in grouped_lines3 if line.startswith("TradeCommType") == False]

test[0:10] #note the removal of the TradeCommType line



['Symbol   Side    Cxl      Qty    Price       Bid        Ask        T-DatS-DatTradeID      TradeTiSS      Exbkr',
 'CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866096       04']

In [82]:
#testing Symbol as well

test = [line for line in test if line.startswith("Symbol") == False]

test[0:10] #note the removal of the Symbol line

['CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866096       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS']

In [87]:
#reassign to grouped_lines3

grouped_lines3 = test

Following the completion of this step, all remaing lines contain only data. However, there is still the problem of data being spread across multiple lines. Folowing the established plan, we will make use of the idea that lines never co-mix and that they always start with a series of 1-4 capital characters followed by at least one space (a corporate ticker) as a way to gauge how many lines one must parse to get a single "line" that contains all relevant data.

In [88]:
#build a regular expression that matches tickers at the start of a string

linestart = re.compile('^[A-Z][A-Z]?[A-Z]?[A-Z]?\s')

#make a new list for this part

grouped_lines4 = grouped_lines3


In [84]:
#put it through its paces

#should match these
print(bool(linestart.match("CERS ")))
print(bool(linestart.match("VAS ")))
print(bool(linestart.match("ZI ")))
print(bool(linestart.match("P ")))

print("")

#not these

print(bool(linestart.match("CERS")))
print(bool(linestart.match("VAS")))
print(bool(linestart.match("ZI")))
print(bool(linestart.match("P")))
print(bool(linestart.match("as")))
print(bool(linestart.match("2.60000")))
print(bool(linestart.match(" CERS")))
print(bool(linestart.match("P")))
print(bool(linestart.match("pA ")))
print(bool(linestart.match("\tFRED")))
print(bool(linestart.match("ZIASE")))
print(bool(linestart.match(".")))

True
True
True
True

False
False
False
False
False
False
False
False
False
False
False
False


In [89]:
#now, test it on actual lines

grouped_lines4[0:10]


['CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS',
 '/17  /17  256945       00',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866096       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS']

In [93]:
#should match first line and 4th line

print(bool(linestart.match(grouped_lines4[0])))
print(bool(linestart.match(grouped_lines4[3])))


print()

#not lines 2 or 3

print(bool(linestart.match(grouped_lines4[1])))
print(bool(linestart.match(grouped_lines4[2])))

True
True

False
False


In [94]:
#create a generic peel function to grab a layer, and a remaing function to go with it

peelG = lambda list : list[0]

leftover = lambda list : list[1:]



In [97]:
#test it

test = grouped_lines4

print(peelG(test))

print()

print(leftover(test)[0:5])

CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS

['/17  /17  256945       00', '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO', 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX', '/17  /17  866087       04', '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS']


In [99]:
#Create a peel function that will pull off a line if it does not match the previous regex, else return flag

peelCon = lambda list : list[0] if bool(linestart.match(list[0])) != True else "Failed"

In [110]:
#test it

test = grouped_lines4

#should fail
print(peelCon(test))

print()

#should peel
test = leftover(test)

print(peelCon(test))

Failed

/17  /17  256945       00


In [111]:
#next, we need to build a basic loop that will peel 1 complete line



test = grouped_lines4

#set up the first line
output = peelG(test)

test = leftover(test)

#pell the rest
while peelCon(test) != "Failed":

    output = output + " " + peelCon(test)
    test = leftover(test)


print(output)


CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS /17  /17  256945       00 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO


In [114]:
#if it worked correctly, the next line will be the start of a new line, as seen here

test[0:5]

['CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX',
 '/17  /17  866087       04',
 '0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS',
 '854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX']

In [118]:
#next, we need to wrap that in a different loop that kicks it off for every line

#test the first 50 lines, keeping in mind the last line printed liely won't be complete
test = grouped_lines4[0:50]

#while there are lines to pull
while test != []:

    #pull off starting line
    output = peelG(test)
    test = leftover(test)

    #after that, pull off any remaing lines, assuming there are any
    while test != [] and peelCon(test) != "Failed" :

        output = output + " " + peelCon(test)
        test = leftover(test)

    print(output)


CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS /17  /17  256945       00 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO
CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX /17  /17  866087       04 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300
CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX /17  /17  866096       04 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300
CERS     SS      NEW          100   2.350000    2.350000   2.36000006/1406/191706149900003 10:20:ContraSSFREX /17  /17  869780       23 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300
CERS     SS      NEW          100   2.350000    2.350000   2.36000006/1406/191706149900003 10:22:ContraSSFREX /17  /17  892888       29 0704 PerUnit   TERM  0.010000     AAB

In [121]:
#now, save outputs  in a list, rather then printing them


#test the first 50 lines, keeping in mind the last line printed liely won't be complete
test = grouped_lines4[0:50]

datalines = []

#while there are lines
while test != []:

    #pull off starting line
    output = peelG(test)
    test = leftover(test)

    #after that, pull off any remaing lines, assuming there are any
    while test != [] and peelCon(test) != "Failed" :

        output = output + " " + peelCon(test)
        test = leftover(test)

    datalines.append(output)

In [122]:
#check the list

datalines[0:10]

['CERS     SS      NEW        2,756   2.400000    2.340000   2.45000006/1406/191706149900003 09:30:CustSS /17  /17  256945       00 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=GSCO',
 'CERS     SS      NEW          100   2.36000018422.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX /17  /17  866087       04 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          200   2.350000    2.360000   2.37000006/1406/191706149900003 10:20:ContraSSFREX /17  /17  866096       04 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          100   2.350000    2.350000   2.36000006/1406/191706149900003 10:20:ContraSSFREX /17  /17  869780       23 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          100   2.350000    2.350000   2.36000006/1406/191706149900003 10:22:ContraSSFREX /17  /17  892888       29 0704 PerUnit   TERM

In [125]:
#now, use the logic to consolidate the entire dataset

#test the first 50 lines, keeping in mind the last line printed liely won't be complete
test = grouped_lines4

datalines = []

#while there are lines
while test != []:

    #pull off starting line
    output = peelG(test)
    test = leftover(test)

    #after that, pull off any remaing lines, assuming there are any
    while test != [] and peelCon(test) != "Failed" :

        output = output + " " + peelCon(test)
        test = leftover(test)

    datalines.append(output)

In [126]:
datalines

35750085432.350000   2.36000006/1406/191706149900006 15:06:ContraSSDTTX /17  /17  364799       32 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          100   2.355000    2.350000   2.36000006/1406/191706149900006 15:07:ContraSSDTTX /17  /17  369938       02 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          100   2.355000    2.350000   2.36000006/1406/191706149900006 15:07:ContraSSDTTX /17  /17  372950       22 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          100   2.355000    2.350000   2.36000006/1406/191706149900006 15:07:ContraSSDTTX /17  /17  372953       22 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS 854300',
 'CERS     SS      NEW          100   2.355000    2.350000   2.36000006/1406/191706149900006 15:07:ContraSSDTTX 0704 PerUnit   TERM  0.010000     AABBC     1706149900003#STA=CERS /17  

In [128]:
#write it out to .csv to inspect

#make it a string
datalines_string = '\n'.join(datalines)

#commented out for safety
#with open('project_1_part_2_output','w') as outfile:
 #   outfile.write(datalines_string)


The original document listed 913 pulls on the end, and after all of this wrangling we are left with 913 singular rows, so no rows were lost. After carefully inspecting the output csv, all rows appear to be pulled properly, though some are inherently dirty.